toolchain/parse/handle_paren_expr.cpp

```diff
@@ -71,14 +78,19 @@ auto HandleExprAfterOpenParenFinish(Context& context) -> void {
 
   // We found a comma, so switch parent state to tuple handling.
   auto finish_state = context.PopState();
-  CARBON_CHECK(finish_state.kind == StateKind::ParenExprFinish,
+  CARBON_CHECK(finish_state.kind == StateKind::ParenExprFinishAsRef ||
+                   finish_state.kind == StateKind::ParenExprFinishAsRegular,
                "Unexpected parent state, found: {0}", finish_state.kind);
   context.PushState(finish_state, StateKind::TupleLiteralFinish);
 
   // If the comma is not immediately followed by a close paren, push handlers
   // for the next tuple element.
   if (list_token_kind != Context::ListTokenKind::CommaClose) {
     context.PushState(state, StateKind::TupleLiteralElementFinish);
```

Comment:
Like for calls, maybe factor out a function to do these pushes and `ref` handling?

Reply:
Done, along with a series of further simplifications that this exposed.